common/flatpak-transaction.c

```diff
@@ -4062,6 +4092,272 @@ flatpak_transaction_normalize_ops (FlatpakTransaction *self)
+      if (runtime_ref != NULL)
+        {
+          g_autofree char *runtime_remote = NULL;
+
+          if (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL)
+            {
+              related = find_related_from_deploy (self, runtime_ref, &runtime_remote);
+            }
+          else
+            {
+              FlatpakTransactionOperation *runtime_op = NULL;
+              runtime_op = flatpak_transaction_get_last_op_for_ref (self, runtime_ref);
+              if (runtime_op != NULL)
+                {
+                  runtime_remote = g_strdup (runtime_op->remote);
+                  if (!op_get_related (self, runtime_op, &related, error))
+                    return FALSE;
+                }
+              else
+                related = find_related_from_deploy (self, runtime_ref, &runtime_remote);
+            }
+
+          if (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL)
+            {
+              g_hash_table_replace (maybe_unused_runtimes, g_strdup (runtime_ref), g_strdup (runtime_remote));
+              g_ptr_array_add (run_after_ops, op);
+            }
+          else
+            g_hash_table_add (newly_used_runtimes, g_strdup (runtime_ref));
+
+          if (!priv->disable_related)
+            {
+              for (i = 0; related && i < related->len; i++)
+                {
+                  FlatpakRelated *rel = g_ptr_array_index (related, i);
+                  if (!rel->delete)
+                    continue;
+
+                  if (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL)
+                    g_hash_table_replace (maybe_unused_runtimes, g_strdup (rel->ref), g_strdup (runtime_remote));
+                  else
+                    g_hash_table_add (newly_used_runtimes, g_strdup (rel->ref));
+                }
+            }
+        }
+
+      /* If the op is changing to a different runtime, the previous one might
+       * now be unused */
+      if (op_type == FLATPAK_TRANSACTION_OPERATION_UPDATE &&
+          dir_ref_is_installed (priv->dir, op->ref, NULL, &deploy_data))
+        {
+          g_autofree char *full_previous_runtime = NULL;
+          g_autofree char *prev_runtime_remote = NULL;
+
+          const char *previous_runtime = flatpak_deploy_data_get_runtime (deploy_data);
+
+          if (previous_runtime == NULL || *previous_runtime == '\0')
+            continue;
+
+          full_previous_runtime = g_strconcat ("runtime/", previous_runtime, NULL);
+          if (g_strcmp0 (full_previous_runtime, runtime_ref) == 0)
+            continue;
+
+          g_clear_pointer (&related, g_ptr_array_unref);
+          related = find_related_from_deploy (self, full_previous_runtime, &prev_runtime_remote);
+
+          g_hash_table_replace (maybe_unused_runtimes,
+                                g_strdup (full_previous_runtime),
+                                g_strdup (prev_runtime_remote));
+
+          g_ptr_array_add (to_be_excluded, op->ref);
```

Comment:
Is this really right? If we're updating gedit and the new one requires a new runtime, should we really ignore the old one completely? That means that e.g. the gedit locale is deemed unused, is it not?

Reply:
You're right I think the gedit locale would be considered unused in the call to `list_unused_refs_with_options()` in `prune_maybe_unused_refs()`, but we only use that set to prune `maybe_unused_runtimes` and the gedit locale would not be in `maybe_unused_runtimes`, so it wouldn't be actually uninstalled.

I think the questions are:
1) Is this code responsible for finding unused related refs of apps or only unused proper runtimes (and their related refs)?
2) Is it possible for an app update to cause an installed related ref, e.g. `org.gimp.GIMP.Plugin.Fourier` to become unused, such as by renaming the defined extension point `org.gimp.GIMP.Plugin`?

Not 100% sure on those, might need to do more digging.

Reply:
Its certainly possible for an app update to make a related ref unused (like dropping an extension point) and we should ideally handle that, and while this is less common than for runtimes to go eol i think it seems useful to do.

I wonder if the current API of excluding refs in list_unused is just too simple. What if we instead could inject the new dependency information into the actual "find used refs" algorithm. 

So, suppose we could pass a hashtable of ref -> metadata information (where metadata=NULL means the ref has doesn't exist). We already have all these metadatas from the resolve stage. Then during the recursion over all the already installed app refs whenever we try to find dependencies we first look at the hashtable for the metadata rather than getting it from the deploy file.

Additionally we have to add any newly installed apps to the initial list of app refs to scan for used refs.

This way we get the full new dependency graph and can do a regular GC on it rather than trying to second-guess what list_unused will do. I think it will make things a lot more robust.

Reply:
Yeah that does sound like a better implementation.